doc/modules/array_api.rst

```diff
@@ -178,21 +178,82 @@ Tools
 Coverage is expected to grow over time. Please follow the dedicated `meta-issue on GitHub
 <https://github.com/scikit-learn/scikit-learn/issues/22352>`_ to track progress.
 
-Type of return values and fitted attributes
--------------------------------------------
+Input and output array type handling
+------------------------------------
 
-When calling functions or methods with Array API compatible inputs, the
+Estimators and scoring functions are able to accept input arrays
+from different array libraries and/or devices. When mixed input arrays are
+passed, scikit-learn silently converted to make them all consistent.
+
+For estimators, the rule is **"everything follows `X`"** - mixed array inputs are
+converted so they are all match the array library and device of `X`.
+For scoring functions the rule is **"everything follows `y_pred`"** - mixed array
+inputs are converted so they are all match the array library and device of `y_pred`.
+
+When a function or method has been called with Array API compatible inputs, the
 convention is to return array values of the same array container type and
 device as the input data.
 
-Similarly, when an estimator is fitted with Array API compatible inputs, the
-fitted attributes will be arrays from the same library as the input and stored
-on the same device. The `predict` and `transform` method subsequently expect
+For details and reasons behind this choice, see below.
+
+Estimators
+~~~~~~~~~~
+
+When an estimator is fitted with an Array API compatible `X`, all other
+array inputs, including constructor arguments, (e.g., `y`, `sample_weight`,
+`weights_init`) will be silently converted to match the array library and
+device of `X`, if they do not already.
+This allows estimators to accept mixed input types, enabling `X` to be moved
+to a different device within a pipeline, without explicitly moving `y`.
+Note that scikit-learn pipelines do not allow transformation of `y` (to avoid
+:ref:`leakage <data_leakage>`).
+
+Take for example a pipeline where `X` and `y` both start on CPU, and go through
+the following three steps:
+
+* :class:`~sklearn.preprocessing.TargetEncoder`, which will transform categorial
+  `X` but also requires `y`, meaning both `X` and `y` need to be on CPU.
+* :class:`~sklearn.preprocessing.FunctionTransformer`, with the function
+  `partial(torch.asarray, device="cuda")`, which moves `X` to GPU, to improve
+  performance in the next step.
+* :class:`~sklearn.linear_model.Ridge`, whose performance can be improved when
+  passed arrays on GPU, as it only takes numerical inputs.
```

Comment:
"on GPU", "on a GPU", "on the GPU"? Not sure which one is right/sounds best (same applies to "on CPU" above) - do you have a preference?